src/BizHawk.Client.Common/input/HostInputAdapter.cs

```diff
@@ -20,6 +20,8 @@ public interface IHostInputAdapter
 
 		IEnumerable<KeyEvent> ProcessHostKeyboards();
 
+		void ProcessVibrationAll(IReadOnlyCollection<(string name, int strength)> haptics);
```

Comment:
The way you've done this means the code is duplicated between OpenTK and DirectInput. (It also seems to merge all channels into one, sending the result to all gamepads?) Try adding a parameter to `IHostInputAdapter.ProcessHostGamepads` instead.

Reply:
While I agree that some code is duplicated, unfortunately, that's simply how these two classes are setup.

DirectInput and OpenTK duplicate code for Buttons and Axis controls too.